internal/controller/template_controller.go

```diff
@@ -439,24 +436,106 @@ func processTemplateResources(
 			log.Error(err, "unable to convert YAML to JSON")
 			return err
 		}
+
 		unstructuredObj := &unstructured.Unstructured{}
 		err = unstructuredObj.UnmarshalJSON(jsonData)
 		if err != nil {
 			log.Error(err, "unable to unmarshal JSON")
 			return err
 		}
-		context := context.Background()
-		_, err = k8sClient.CreateOrUpdate(context, unstructuredObj)
 
+		// Build resource key
+		key := resourceKey(unstructuredObj)
+
+		// Add to desired resources
+		desiredResources[key] = unstructuredObj
+
+		// Create ResourceIdentifier
+		ri := ResourceIdentifier{
+			Group:     unstructuredObj.GroupVersionKind().Group,
+			Version:   unstructuredObj.GroupVersionKind().Version,
+			Kind:      unstructuredObj.GroupVersionKind().Kind,
+			Namespace: unstructuredObj.GetNamespace(),
+			Name:      unstructuredObj.GetName(),
+		}
+
+		desiredResourceIdentifiers = append(desiredResourceIdentifiers, ri)
+	}
+
+	// Retrieve existing resources generated by this template
+	wm.mu.Lock()
+	existingResourceIdentifiers := wm.generatedResources[template.Name]
+	wm.mu.Unlock()
+
+	// Build map of existing resources
+	existingResources := make(map[string]ResourceIdentifier)
+	for _, ri := range existingResourceIdentifiers {
+		key := resourceKeyFromRI(ri)
+		existingResources[key] = ri
+	}
+
+	var resourcesToDelete []ResourceIdentifier
+	for key, ri := range existingResources {
+
+		fmt.Println("Checking if resource exists", "key", key, "resource", ri)
+		if _, exists := desiredResources[key]; !exists {
+			resourcesToDelete = append(resourcesToDelete, ri)
+			fmt.Println("Resource to deleteeee", "resource", ri)
+		}
+	}
+
+	// Delete unwanted resources
+	for _, ri := range resourcesToDelete {
+		err := deleteResource(ri, k8sClient, log)
+		if err != nil {
+			log.Error(err, "unable to delete resource", "resource", ri)
+			return err
+		}
+	}
+
+	// Apply desired resources (create or update)
+	for _, unstructuredObj := range desiredResources {
+		_, err = k8sClient.CreateOrUpdate(context.Background(), unstructuredObj)
 		if err != nil {
-			if strings.Contains(err.Error(), "already exists") {
-				log.Info("resource already exists", "resource", unstructuredObj.GetName())
-			} else {
-				log.Error(err, "unable to create resource")
-				return err
-			}
+			log.Error(err, "unable to create or update resource", "resource", resourceKey(unstructuredObj))
+			return err
 		}
 	}
 
+	// Update the generatedResources in WatchManager
+	wm.mu.Lock()
+	wm.generatedResources[template.Name] = desiredResourceIdentifiers
```

Comment:
Instead of directly updating it (and dealing with lock/unlock), how about encapsulating this in a method on `WatchManager`?

Reply:
Sure. Actually my initial plan was to first have this and then during refactoring create a new struct for a template itself and have it as a list inside this struct. We have now something like this

		templates:          make(map[string]map[schema.GroupVersionKind]struct{}),
		generatedResources: make(map[string][]ResourceIdentifier)
and they all share a same lock. I was thinking of having a struct for each template so they can also have their own lock instead of using a shared lock which is not very efficient. That's why for now I just wanted to make it work until refactoring. But I can do that since  we need to have this kind of encapsulation anyway during refactoring. So let's do it now

Reply:
done